numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
The pattern I'm finding is `return (needs_api && PyErr_Occurred()) ? -1 : 0;` though - so are you sure that:

* We need to catch the error mid-loop, rather than wait till the end
* We need to catch anything at all if `!needs_api`